File: jrt_test/src/jrt-test/simpleserver/simpleserver.cpp
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

#include <vespa/fnet/frt/supervisor.h>
#include <vespa/fnet/frt/rpcrequest.h>
#include <vespa/fnet/transport.h>
#include <vespa/fastos/app.h>

class Server : public FRT_Invokable
{
private:
    Server(const Server &);
    Server &operator=(const Server &);

public:
    Server(FRT_Supervisor *s)
    {
        FRT_ReflectionBuilder rb(s);
        //---------------------------------------------------------------------
        rb.DefineMethod("inc", "i", "i",
                        FRT_METHOD(Server::rpc_inc), this);
        rb.MethodDesc("Increase an integer value");
        rb.ParamDesc("value", "initial value");
        rb.ReturnDesc("result", "value + 1");
        //---------------------------------------------------------------------
        rb.DefineMethod("blob", "x", "x",
                        FRT_METHOD(Server::rpc_blob), this);
        rb.MethodDesc("Send a copy of a blob back to the client");
        rb.ParamDesc("blob", "the original blob");
        rb.ReturnDesc("blob", "a copy of the original blob");
        //---------------------------------------------------------------------
        rb.DefineMethod("test", "iib", "i",
                        FRT_METHOD(Server::rpc_test), this);
        rb.MethodDesc("Magic test method");
        rb.ParamDesc("value", "the value");
        rb.ParamDesc("error", "error code to set");
        rb.ParamDesc("extra", "if not 0, add an extra return value");
        rb.ReturnDesc("value", "the value");
        //---------------------------------------------------------------------
    }

    void rpc_inc(FRT_RPCRequest *req)
    {
        req->GetReturn()->AddInt32(req->GetParams()->GetValue(0)._intval32 + 1);
    }

    void rpc_blob(FRT_RPCRequest *req)
    {
        req->GetReturn()->AddData(req->GetParams()->GetValue(0)._data._buf,
                                  req->GetParams()->GetValue(0)._data._len);
    }

    void rpc_test(FRT_RPCRequest *req)
    {
	int value = req->GetParams()->GetValue(0)._intval32;
	int error = req->GetParams()->GetValue(1)._intval32;
	int extra = req->GetParams()->GetValue(2)._intval8;

	req->GetReturn()->AddInt32(value);
	if (extra != 0) {
	    req->GetReturn()->AddInt32(value);
	}
	if (error != 0) {
	    req->SetError(error, "Custom error");
	}
    }
};


class App : public FastOS_Application
{
public:
    int Main() override;
};


int
App::Main()
{
    if (_argc < 2) {
        printf("usage: %s <listenspec>\n", _argv[0]);
        return 1;
    }
    fnet::frt::StandaloneFRT frt;
    Server server(&frt.supervisor());
    frt.supervisor().Listen(_argv[1]);
    frt.supervisor().GetTransport()->WaitFinished();
    return 0;
}


int
main(int argc, char **argv)
{
    App myapp;
    return myapp.Entry(argc, argv);
}

File: jrt_test/src/java/SimpleServer.java
// Copyright Vespa.ai. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

import com.yahoo.jrt.*;

public class SimpleServer {

    public void rpc_inc(Request req) {
	req.returnValues().add(new Int32Value(req.parameters().get(0).asInt32()
					      + 1));
    }

    public void rpc_echo(Request req) {
	for (int i = 0; i < req.parameters().size(); i++) {
	    req.returnValues().add(req.parameters().get(i));
	}
    }

    public void rpc_test(Request req) {
	int value = req.parameters().get(0).asInt32();
	int error = req.parameters().get(1).asInt32();
	int extra = req.parameters().get(2).asInt8();

	req.returnValues().add(new Int32Value(value));
	if (extra != 0) {
	    req.returnValues().add(new Int32Value(value));	
	}
	if (error != 0) {
	    req.setError(error, "Custom error");
	}
    }

    public static void main(String[] args) {
	if (args.length != 1) {
	    System.err.println("usage: SimpleServer <spec>");
	    System.exit(1);
	}
	Supervisor orb = new Supervisor(new Transport());
	SimpleServer handler = new SimpleServer();
	orb.addMethod(new Method("inc", "i", "i", handler::rpc_inc));
	orb.addMethod(new Method("echo", "*", "*", handler::rpc_echo));
	orb.addMethod(new Method("test", "iib", "i", handler::rpc_test));
	try {
	    orb.listen(new Spec(args[0]));
	} catch (ListenFailedException e) {
	    System.err.println("could not listen at " + args[0]);
	    System.exit(1);	
	}
	orb.transport().join();
    }
}
